File: 1201/Heap/heapSort.java
import java.util.ArrayList;
public class heapSort{
    
    public static void downHeapify(int[] arr,int parIdx,int ei,boolean isMax){ //logn
        int maxidx=parIdx;
        int lci=2*parIdx+1;
        int rci=2*parIdx+2;

        if(lci<=ei && compareTo(arr,lci,maxidx,isMax)) maxidx=lci;
        if(rci<=ei&& compareTo(arr,rci,maxidx,isMax)) maxidx=rci;
        
        if(parIdx!=maxidx){
            swap(arr,maxidx,parIdx);
            downHeapify(arr,maxidx,ei,isMax);
        }
    }

    public static void swap(int[] arr,int a,int b){
        int temp=arr[a];
        arr[a]=arr[b];
        arr[b]=temp;
    }

    public static boolean compareTo(int[] arr,int ci,int pi,boolean isMax){
        if(isMax) return arr[ci]>arr[pi];
        else return arr[ci]<arr[pi];
    }

    public static void main(String[] args){
        int[] arr={6,8,2,-4,18,4,10,9,8,9,16,15,13,11,-9,-10};
        int ei=arr.length-1;
        boolean isMax=true;
        for(int i=ei;i>=0;i--){
            downHeapify(arr,i,ei,isMax);
        }

        while(ei!=0){
            swap(arr,0,ei);
            ei--;
            downHeapify(arr,0,ei,isMax);
        }

        for(int ele: arr) System.out.print(ele+" ");


     }



}

File: 2019/levelUpAugBatch/lecture007_HashMapHeap/Heap/heapSort.cpp
#include <iostream>
#include <vector>

using namespace std;

bool compareTo(int a, int b, vector<int> &arr, bool isMax)
{
    if (isMax)
        return arr[a] > arr[b];
    else
        return arr[b] > arr[a];
}

void downHeapify(int pi, int n, vector<int> &arr, bool isMax)
{
    int maxIdx = pi;
    int lci = 2 * pi + 1;
    int rci = 2 * pi + 2;

    if (lci <= n && compareTo(lci, maxIdx, arr, isMax))
        maxIdx = lci;
    if (rci <= n && compareTo(rci, maxIdx, arr, isMax))
        maxIdx = rci;

    if (maxIdx != pi)
    {
        swap(arr[maxIdx], arr[pi]);
        downHeapify(maxIdx, n, arr, isMax);
    }
}

void heapSort()
{
    vector<int> arr = {10, 20, 30, -2, -3, -4, 5, 6, 7, 8, 9, 22, 11, 13};
    bool isMax = false;
    int n = arr.size() - 1;
    for (int i = n; i >= 0; i--)
        downHeapify(i, n, arr, isMax);

    for (int i = n; i >= 0; i--)
    {
        swap(arr[0], arr[n--]);
        downHeapify(0, n, arr, isMax);
    }

    for (int d : arr)
        cout << d << " ";
}
int main()
{
    heapSort();
    return 0;
}